modules/catalog/src/main/java/org/apache/ignite/internal/catalog/ClockWaiter.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.internal.util.PendingComparableValuesTracker;
+import org.apache.ignite.internal.util.TrackerClosedException;
+import org.apache.ignite.lang.NodeStoppingException;
+
+/**
+ * Allows to wait for the supplied clock to reach a required timesdtamp. It only uses the clock itself,
+ * no SafeTime mechanisms are involved.
+ */
+public class ClockWaiter implements IgniteComponent {
+    private static final IgniteLogger LOG = Loggers.forClass(ClockWaiter.class);
+
+    private final String nodeName;
+    private final HybridClock clock;
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean stopGuard = new AtomicBoolean(false);
+
+    private final PendingComparableValuesTracker<HybridTimestamp, Void> nowTracker = new PendingComparableValuesTracker<>(
+            new HybridTimestamp(1, 0)
+    );
+
+    private final ClockUpdateListener updateListener = this::onUpdate;
+
+    private volatile ScheduledExecutorService scheduler;
+
+    public ClockWaiter(String nodeName, HybridClock clock) {
+        this.nodeName = nodeName;
+        this.clock = clock;
+    }
+
+    @Override
+    public void start() {
+        clock.addUpdateListener(updateListener);
+
+        scheduler = Executors.newSingleThreadScheduledExecutor(new NamedThreadFactory(nodeName + "-clock-waiter", LOG));
+    }
+
+    @Override
+    public void stop() throws Exception {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        clock.removeUpdateListener(updateListener);
+
+        nowTracker.close();
+
+        scheduler.shutdownNow();
+        scheduler.awaitTermination(10, TimeUnit.SECONDS);
+    }
+
+    private void onUpdate(HybridTimestamp newTs) {
+        nowTracker.update(newTs, null);
+    }
+
+    /**
+     * Wait for the clock to reach the given timestamp.
+     *
+     * @param targetTimestamp Timestamp to wait for.
+     * @return A future that completes when the timestamp is reached by the clock's time.
+     */
+    public CompletableFuture<Void> waitFor(HybridTimestamp targetTimestamp) {
+        if (!busyLock.enterBusy()) {
+            return failedFuture(new NodeStoppingException());
+        }
+
+        try {
+            return doWaitFor(targetTimestamp);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    private CompletableFuture<Void> doWaitFor(HybridTimestamp targetTimestamp) {
+        CompletableFuture<Void> future = nowTracker.waitFor(targetTimestamp);
+
+        AtomicReference<ScheduledFuture<?>> scheduledFutureRef = new AtomicReference<>();
```

Comment:
We need a container for a `ScheduledFuture` because it's going to be used in a lambda, so it must be effectively final; without a container it would not compile.

Also, we need the container to be thread-safe (as the future completion might be triggered from another thread), hence `AtomicReference` seems to be an ideal choice.

Reply:
> it's going to be used in a lambda, so it must be effectively final; without a container it would not compile.

It is possible to make `scheduledFuture` effectively final without using a container, you just need to assign `null`s explicitly

> Also, we need the container to be thread-safe (as the future completion might be triggered from another thread), hence AtomicReference seems to be an ideal choice.

This is a local effectively final variable, is it even possible to have races here?